qubes/ext/core_features.py

```diff
@@ -104,7 +104,90 @@ async def qubes_features_request(self, vm, event, untrusted_features):
             untrusted_value = untrusted_features["qubes-agent-version"]
             if _version_re.fullmatch(untrusted_value):
                 vm.features["qubes-agent-version"] = untrusted_value
+
+        # handle boot mode advertisement
+        old_bootmode_info = {}
+        for feature_key, feature_val in vm.features.items():
+            if feature_key.startswith(
+                "boot-mode.kernelopts."
+            ) or feature_key.startswith("boot-mode.name"):
+                old_bootmode_info[feature_key] = feature_val
+        new_bootmode_info = {}
+        untrusted_sanitized_features = {}
+        for (
+            untrusted_feature_key,
+            untrusted_feature_value,
+        ) in untrusted_features.items():
+            if not all(c in string.printable for c in untrusted_feature_value):
```

Comment:
First of all, `untrusted_features` is already filtered for quite limited char set - letters, digits and `"-._ "`: https://github.com/QubesOS/qubes-core-admin/blob/main/qubes/api/misc.py#L61
But if you want to apply other limits anyway, you can probably do that in the next loop, after checking if the feature is relevant for this part of the code; there is no need to sanitize features that are irrelevant for this loop anyway.
Or if you hate to duplicate this line (for both kernelopts and name), then maybe add filtering on `.startswith("boot-mode.")`?

Reply:
If this is already filtered, I don't think filtering it a second time is needed at all. I'll just remove it.

Reply:
Removed.